File: hphp/hack/src/hhbc/decl_vars.ml
(*
 * Copyright (c) 2017, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the "hack" directory of this source tree.
 *
 *)
open Core_kernel
module T = Aast
module ULS = Unique_list_string
module SN = Naming_special_names

type bare_this_usage =
  (* $this appear as expression *)
  | Bare_this
  (* bare $this appear as possible targetfor assignment
  - $this as function argument
  - $this as target of & operator
  - $this as reference in catch clause *)
  | Bare_this_as_ref

type decl_vars_state = {
  (* set of locals used inside the functions *)
  dvs_locals: ULS.t;
  (* does function uses bare form of $this *)
  dvs_bare_this: bare_this_usage option;
}

let with_local name s = { s with dvs_locals = ULS.add s.dvs_locals name }

let with_this barethis s =
  let new_bare_this =
    match (s.dvs_bare_this, barethis) with
    | (_, Bare_this_as_ref) -> Some Bare_this_as_ref
    | (None, Bare_this) -> Some Bare_this
    | (u, _) -> u
  in
  if s.dvs_bare_this = new_bare_this then
    s
  else
    {
      dvs_bare_this = new_bare_this;
      dvs_locals = ULS.add s.dvs_locals SN.SpecialIdents.this;
    }

let dvs_empty = { dvs_locals = ULS.empty; dvs_bare_this = None }

(* Add a local to the accumulated list. Don't add if it's $GLOBALS or
 * the pipe variable $$. If it's $this, add it, and if this variable appears
 * "bare" (because bareparam=true), remember for needs_local_this *)
let add_local ~barethis s (_, name) =
  if
    name = SN.Superglobals.globals
    || name = SN.SpecialIdents.dollardollar
    || SN.SpecialIdents.is_tmp_var name
  then
    s
  else if name = SN.SpecialIdents.this then
    with_this barethis s
  else
    with_local name s

let on_class_get visitor (_, cid_) cge ~is_call_target =
  match cid_ with
  | T.CIparent
  | T.CIself
  | T.CIstatic ->
    failwith "Expects CIexpr as class_id on aast where expr was on ast"
  | T.CI _sid ->
    failwith "Expects CIexpr as class_id on aast where expr was on ast"
  | T.CIexpr e ->
    let _ = visitor#on_expr () e in
    begin
      match cge with
      | T.CGstring pstr ->
        (* TODO: Thomas: For this to match correctly, we need to adjust ast_to_nast
         * because it does not make a distinction between ID and Lvar, which is needed here
         *)
        if is_call_target then
          visitor#set_state (add_local ~barethis:Bare_this visitor#state pstr)
        else
          ()
      | T.CGexpr e2 -> visitor#on_expr () e2
    end

let declvar_visitor explicit_use_set_opt is_in_static_method is_closure_body =
  let state = ref dvs_empty in
  object (self)
    inherit [_] Aast.iter as super

    method! on_stmt_ _ s =
      match s with
      | Aast.Try (body, catch_list, finally) ->
        self#on_block () body;
        List.iter catch_list (fun (_exn_name, (pos, id), catch_body) ->
            state :=
              add_local
                ~barethis:Bare_this_as_ref
                !state
                (pos, Local_id.get_name id);
            self#on_block () catch_body);
        self#on_block () finally
      | _ -> super#on_stmt_ () s

    method! on_expr_ _ e =
      match e with
      | Aast.Obj_get (receiver_e, prop_e, _) ->
        (match snd receiver_e with
        | Aast.Lvar (_, id) when Local_id.get_name id = "$this" ->
          if is_in_static_method && not is_closure_body then
            ()
          else
            state := with_local (Local_id.get_name id) !state
        | _ -> super#on_expr () receiver_e);
        (match snd prop_e with
        (* Only add if it is a variable *)
        | Aast.Lvar (pos, id) ->
          state :=
            add_local ~barethis:Bare_this !state (pos, Local_id.get_name id)
        | _ -> super#on_expr () prop_e)
      | Aast.Binop (binop, e1, e2) ->
        (match (binop, e2) with
        | (Ast_defs.Eq _, (_, Aast.Await _))
        | (Ast_defs.Eq _, (_, Aast.Yield _))
        | (Ast_defs.Eq _, (_, Aast.Yield_from _)) ->
          (* Visit e2 before e1. The ordering of declvars in async
                expressions matters to HHVM. See D5674623. *)
          self#on_expr () e2;
          self#on_expr () e1
        | _ -> super#on_expr_ () e)
      | Aast.Lvar (pos, id) ->
        state := add_local ~barethis:Bare_this !state (pos, Local_id.get_name id)
      | Aast.Class_get (cid, expr) ->
        on_class_get self cid expr ~is_call_target:false
      | Aast.Lfun _ ->
        (* For an Lfun, we don't want to recurse, because it's a separate scope. *)
        ()
      | Aast.Efun (fun_, use_list) ->
        (* at this point AST is already rewritten so use lists on EFun nodes
            contain list of captured variables. However if use list was initially absent
            it is not correct to traverse such nodes to collect locals because it will impact
            the order of locals in generated .declvars section:
            // .declvars $a, $c, $b
            $a = () => { $b = 1 };
            $c = 1;
            $b = 2;
            // .declvars $a, $b, $c
            $a = function () use ($b) => { $b = 1 };
            $c = 1;
            $b = 2;

            'explicit_use_set' is used to in order to avoid synthesized use list *)
        let fn_name = snd fun_.Aast.f_name in
        let has_use_list =
          Option.value_map explicit_use_set_opt ~default:false ~f:(fun s ->
              SSet.mem fn_name s)
        in
        if has_use_list then
          List.iter use_list (fun (pos, id) ->
              state :=
                add_local ~barethis:Bare_this !state (pos, Local_id.get_name id))
        else
          ()
      | Aast.Call (_, func_e, _, pos_args, unpacked_arg) ->
        (match func_e with
        | (_, Aast.Id (pos, call_name))
          when call_name = SN.EmitterSpecialFunctions.set_frame_metadata ->
          state := add_local ~barethis:Bare_this !state (pos, "$86metadata")
        | _ -> ());
        let barethis =
          match func_e with
          | (_, Aast.Id (_, name))
            when name = SN.PseudoFunctions.isset
                 || "\\" ^ name = SN.PseudoFunctions.echo ->
            Bare_this
          | _ -> Bare_this_as_ref
        in
        let on_arg e =
          match e with
          (* Only add $this to locals if it's bare *)
          | (_, Aast.Lvar (_, id)) when Local_id.get_name id = "$this" ->
            state := with_this barethis !state
          | _ -> self#on_expr () e
        in
        (match snd func_e with
        | Aast.Class_get (id, prop) ->
          on_class_get self id prop ~is_call_target:true
        | _ -> self#on_expr () func_e);
        List.iter pos_args on_arg;
        Option.iter unpacked_arg on_arg
      | Aast.New (_, _, exprs1, expr2, _) ->
        let add_bare_expr expr =
          match expr with
          | (_, Aast.Lvar (_, id)) when Local_id.get_name id = "$this" ->
            state := with_this Bare_this_as_ref !state
          | _ -> self#on_expr () expr
        in
        List.iter exprs1 add_bare_expr;
        Option.iter expr2 add_bare_expr
      | _ -> super#on_expr_ () e

    method! on_class_ _ _ = ()

    method! on_fun_def _ _ = ()

    method state = !state

    method set_state new_state = state := new_state
  end

let uls_from_ast
    ~is_closure_body
    ~has_this
    ~params
    ~is_toplevel
    ~is_in_static_method
    ~get_param_name
    ~get_param_default_value
    ~explicit_use_set_opt
    (b : Tast.program) =
  let visitor =
    declvar_visitor explicit_use_set_opt is_in_static_method is_closure_body
  in
  (* pull variables used in default values *)
  List.iter params (fun p ->
      match get_param_default_value p with
      | Some e -> visitor#on_expr () e
      | _ -> ());
  visitor#on_program () b;
  let state = visitor#state in
  let needs_local_this =
    state.dvs_bare_this = Some Bare_this_as_ref || is_in_static_method
  in
  let param_names =
    List.fold_left params ~init:ULS.empty ~f:(fun l p ->
        ULS.add l @@ get_param_name p)
  in
  let decl_vars = ULS.diff state.dvs_locals param_names in
  let decl_vars =
    if needs_local_this || is_closure_body || (not has_this) || is_toplevel then
      decl_vars
    else
      ULS.remove "$this" decl_vars
  in
  (needs_local_this && has_this, decl_vars)

(* See decl_vars.mli for details *)
let from_ast
    ~is_closure_body
    ~has_this
    ~params
    ~is_toplevel
    ~is_in_static_method
    ~explicit_use_set
    (b : Tast.program) =
  let (needs_local_this, decl_vars) =
    uls_from_ast
      ~is_closure_body
      ~has_this
      ~params
      ~is_toplevel
      ~is_in_static_method
      ~get_param_name:Hhas_param.name
      ~get_param_default_value:(fun p ->
        Option.map (Hhas_param.default_value p) ~f:snd)
      ~explicit_use_set_opt:(Some explicit_use_set)
      b
  in
  (needs_local_this, ULS.items decl_vars)

let vars_from_ast
    ~is_closure_body
    ~has_this
    ~(params : Tast.fun_param list)
    ~is_toplevel
    ~is_in_static_method
    (b : Tast.program) =
  let (_, decl_vars) =
    uls_from_ast
      ~is_closure_body
      ~has_this
      ~params
      ~is_toplevel
      ~is_in_static_method
      ~get_param_name:(fun p -> p.Aast.param_name)
      ~get_param_default_value:(fun p -> p.Aast.param_expr)
      ~explicit_use_set_opt:None
      b
  in
  ULS.items_set decl_vars

File: hphp/hack/src/hhbc/decl_vars.rs
// Copyright (c) Facebook, Inc. and its affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the "hack" directory of this source tree.

use itertools::Either;
use std::{collections::HashSet, iter::Iterator};

extern crate bitflags;
use bitflags::bitflags;

use hhas_param_rust::HhasParam;
use naming_special_names_rust::{
    emitter_special_functions, pseudo_functions, special_idents, superglobals,
};
use oxidized::{
    aast,
    aast_visitor::{visit, AstParams, Node, Visitor},
    ast::*,
};
use unique_list_rust::UniqueList;

bitflags! {
    pub struct Flags: u8 {
        const IS_CLOSURE_BODY = 1 << 1;
        const HAS_THIS = 1 << 2;
        const IS_TOPLEVEL = 1 << 3;
        const IS_IN_STATIC_METHOD = 1 << 4;
    }
}

pub type ProgramOrStmt<'a> = Either<&'a Program, &'a Block>;

#[derive(PartialEq, Copy, Clone)]
enum BareThisUsage {
    // $this appear as expression
    BareThis,
    // bare $this appear as possible targetfor assignment
    // - $this as function argument
    // - $this as target of & operator
    // - $this as reference in catch clause *)
    BareThisAsRef,
}

use BareThisUsage::*;

struct DeclvarVisitorContext<'a> {
    explicit_use_set_opt: Option<&'a env::SSet>,
    is_in_static_method: bool,
    is_closure_body: bool,
}

struct DeclvarVisitor<'a> {
    // set of locals used inside the functions
    locals: UniqueList<String>,
    // does function uses bare form of $this
    bare_this: Option<BareThisUsage>,
    context: DeclvarVisitorContext<'a>,
}

impl<'a> DeclvarVisitor<'a> {
    fn new(explicit_use_set_opt: Option<&'a env::SSet>, flags: Flags) -> Self {
        Self {
            locals: UniqueList::new(),
            bare_this: None,
            context: DeclvarVisitorContext {
                explicit_use_set_opt,
                is_in_static_method: flags.contains(Flags::IS_IN_STATIC_METHOD),
                is_closure_body: flags.contains(Flags::IS_CLOSURE_BODY),
            },
        }
    }

    fn with_this(&mut self, barethis: BareThisUsage) {
        let new_bare_this = match (self.bare_this, barethis) {
            (_, BareThisAsRef) => Some(BareThisAsRef),
            (None, BareThis) => Some(BareThis),
            (u, _) => u,
        };
        if self.bare_this != new_bare_this {
            self.bare_this = new_bare_this;
            self.locals.add(special_idents::THIS.into())
        }
    }

    fn add_local<S: Into<String> + AsRef<str>>(&mut self, barethis: BareThisUsage, name: S) {
        let name_ref = name.as_ref();
        if name_ref == superglobals::GLOBALS
            || name_ref == special_idents::DOLLAR_DOLLAR
            || special_idents::is_tmp_var(name_ref)
        {
            ()
        } else if name_ref == special_idents::THIS {
            self.with_this(barethis)
        } else {
            self.locals.add(name.into())
        }
    }

    fn on_class_get(
        &mut self,
        cid: &ClassId,
        cge: &ClassGetExpr,
        is_call_target: bool,
    ) -> Result<(), String> {
        use aast::ClassId_::*;
        match &cid.1 {
            CIparent | CIself | CIstatic | CI(_) => {
                Err("Expects CIexpr as class_id on aast where expr was on ast".into())
            }
            CIexpr(e) => {
                self.visit_expr(&mut (), e)?;
                use aast::ClassGetExpr::*;
                match cge {
                    CGstring(pstr) => {
                        // TODO(thomasjiang): For this to match correctly, we need to adjust ast_to_nast
                        // because it does not make a distinction between ID and Lvar, which is needed here
                        if is_call_target {
                            self.add_local(BareThis, &pstr.1)
                        }
                        Ok(())
                    }
                    CGexpr(e2) => self.visit_expr(&mut (), e2),
                }
            }
        }
    }
}

impl<'a> Visitor for DeclvarVisitor<'a> {
    type P = AstParams<(), String>;

    fn object(&mut self) -> &mut dyn Visitor<P = Self::P> {
        self
    }

    fn visit_stmt_(&mut self, env: &mut (), s: &Stmt_) -> Result<(), String> {
        match s {
            Stmt_::Try(x) => {
                let (body, catch_list, finally) = (&x.0, &x.1, &x.2);
                visit(self, env, body)?;
                for Catch(_, id, catch_body) in catch_list {
                    self.add_local(BareThisAsRef, id.name());
                    visit(self, env, catch_body)?;
                }
                visit(self, env, finally)
            }
            _ => s.recurse(env, self),
        }
    }

    fn visit_expr_(&mut self, env: &mut (), e: &Expr_) -> Result<(), String> {
        use aast::Expr_::*;
        match e {
            ObjGet(x) => {
                let (receiver_e, prop_e) = (&x.0, &x.1);
                match &receiver_e.1 {
                    Lvar(id) if id.name() == "$this" => {
                        if self.context.is_in_static_method && !self.context.is_closure_body {
                        } else {
                            self.locals.add(id.name().into())
                        }
                    }
                    _ => receiver_e.recurse(env, self.object())?,
                }
                match &prop_e.1 {
                    Lvar(id) => self.add_local(BareThis, id.name()),
                    _ => prop_e.recurse(env, self.object())?,
                }
                Ok(())
            }

            Binop(x) => {
                let (binop, e1, e2) = (&x.0, &x.1, &x.2);
                match (binop, &e2.1) {
                    (Bop::Eq(_), Await(_))
                    | (Bop::Eq(_), Yield(_))
                    | (Bop::Eq(_), YieldFrom(_)) => {
                        // Visit e2 before e1. The ordering of declvars in async
                        // expressions matters to HHVM. See D5674623.
                        self.visit_expr(env, e2)?;
                        self.visit_expr(env, e1)
                    }
                    _ => e.recurse(env, self.object()),
                }
            }

            Lvar(x) => Ok(self.add_local(BareThis, x.name())),
            ClassGet(x) => self.on_class_get(&x.0, &x.1, false),
            // For an Lfun, we don't want to recurse, because it's a separate scope.
            Lfun(_) => Ok(()),
            Efun(x) => {
                // at this point AST is already rewritten so use lists on EFun nodes
                // contain list of captured variables. However if use list was initially absent
                // it is not correct to traverse such nodes to collect locals because it will impact
                // the order of locals in generated .declvars section:
                // // .declvars $a, $c, $b
                // $a = () => { $b = 1 };
                // $c = 1;
                // $b = 2;
                // // .declvars $a, $b, $c
                // $a = function () use ($b) => { $b = 1 };
                // $c = 1;
                // $b = 2;
                //
                // 'explicit_use_set' is used to in order to avoid synthesized use list
                let (fun_, use_list) = (&x.0, &x.1);
                let fn_name = &fun_.name.1;
                let has_use_list = self
                    .context
                    .explicit_use_set_opt
                    .map(|s| s.contains(fn_name))
                    .unwrap_or(false);
                if has_use_list {
                    for id in use_list {
                        self.add_local(BareThis, id.name())
                    }
                }
                Ok(())
            }
            Call(x) => {
                let (func_e, pos_args, unpacked_arg) = (&x.1, &x.3, &x.4);
                if let Id(x) = &func_e.1 {
                    let call_name = &x.1;
                    if call_name == emitter_special_functions::SET_FRAME_METADATA {
                        self.add_local(BareThis, "$86metadata");
                    }
                }
                let barethis = match &func_e.1 {
                    Id(name)
                        if &name.1 == pseudo_functions::ISSET
                            || &name.1 == pseudo_functions::ECHO_NO_NS =>
                    {
                        BareThis
                    }
                    _ => BareThisAsRef,
                };
                let on_arg = |self_: &mut Self, env: &mut (), x: &Expr| match &x.1 {
                    // Only add $this to locals if it's bare
                    Lvar(id) if &(id.1).1 == "$this" => Ok(self_.with_this(barethis)),
                    _ => self_.visit_expr(env, x),
                };
                match &func_e.1 {
                    ClassGet(x) => {
                        let (id, prop) = (&x.0, &x.1);
                        self.on_class_get(id, prop, true)?
                    }
                    _ => self.visit_expr(env, func_e)?,
                }
                for arg in pos_args {
                    on_arg(self, env, arg)?
                }
                if let Some(arg) = unpacked_arg {
                    on_arg(self, env, arg)?
                }
                Ok(())
            }
            New(x) => {
                let (exprs1, expr2) = (&x.2, &x.3);

                let add_bare_expr = |self_: &mut Self, env: &mut (), expr: &Expr| match &expr.1 {
                    Lvar(x) if &(x.1).1 == "$this" => Ok(self_.with_this(BareThisAsRef)),
                    _ => self_.visit_expr(env, expr),
                };
                for expr in exprs1 {
                    add_bare_expr(self, env, expr)?;
                }
                if let Some(expr) = expr2 {
                    add_bare_expr(self, env, expr)?
                }
                Ok(())
            }
            e => e.recurse(env, self.object()),
        }
    }

    fn visit_class_(&mut self, _: &mut (), _: &Class_) -> Result<(), String> {
        Ok(())
    }
    fn visit_fun_(&mut self, _: &mut (), _: &Fun_) -> Result<(), String> {
        Ok(())
    }
}

fn uls_from_ast<P, F1, F2>(
    params: &[P],
    get_param_name: F1,
    get_param_default_value: F2,
    explicit_use_set_opt: Option<&env::SSet>,
    b: ProgramOrStmt,
    flags: Flags,
) -> Result<(bool, impl Iterator<Item = String>), String>
where
    F1: Fn(&P) -> &str,
    F2: Fn(&P) -> Option<&Expr>,
{
    let mut visitor = DeclvarVisitor::new(explicit_use_set_opt, flags);

    for p in params {
        if let Some(e) = get_param_default_value(p) {
            visitor.visit_expr(&mut (), e)?;
        }
    }
    match b {
        Either::Left(b) => visit(&mut visitor, &mut (), b)?,
        Either::Right(b) => visit(&mut visitor, &mut (), b)?,
    };
    let needs_local_this =
        visitor.bare_this == Some(BareThisAsRef) || flags.contains(Flags::IS_IN_STATIC_METHOD);
    for param in params {
        visitor.locals.remove(get_param_name(param))
    }
    if !(needs_local_this
        || flags.contains(Flags::IS_CLOSURE_BODY)
        || !flags.contains(Flags::HAS_THIS)
        || flags.contains(Flags::IS_TOPLEVEL))
    {
        visitor.locals.remove("$this")
    }
    Ok((
        needs_local_this && flags.contains(Flags::HAS_THIS),
        visitor.locals.into_iter(),
    ))
}

pub fn from_ast(
    params: &[HhasParam],
    body: &Program,
    flags: Flags,
    explicit_use_set: &env::SSet,
) -> Result<(bool, Vec<String>), String> {
    let (needs_local_this, decl_vars) = uls_from_ast(
        params,
        |p| &p.name,
        |p| p.default_value.as_ref().map(|x| &x.1),
        Some(explicit_use_set),
        Either::Left(body),
        flags,
    )?;
    Ok((needs_local_this, decl_vars.collect()))
}

pub fn vars_from_ast(
    params: &[FunParam],
    b: ProgramOrStmt,
    flags: Flags,
) -> Result<HashSet<String>, String> {
    let (_, decl_vars) = uls_from_ast(
        params,
        |p| &p.name,         // get_param_name
        |p| p.expr.as_ref(), // get_param_default_value
        None,                // explicit_use_set_opt
        b,
        flags,
    )?;
    Ok(decl_vars.collect())
}
